### ISSUE TITLE
Null thrustCurve nodes

### ISSUE THREAD

**Issue by PhineasFreak** [NathanKell/ModularFuelSystem#171]
So, i was searching the KSP logs for warnings or errors for another mod and i came across this one from RF:

```
*RFE* ERROR: have curve node but thrustCurve is null!
```

Of course this message is a sample one. I checked all SRM parts that implement thrust curves and they all reported the same thing. The part config reports the correct config module structure though (from MM):

```
MODULE
{
    name = ModuleEngineConfigs
    configuration = Castor-I
    modded = false
    type = ModuleEnginesRF
    CONFIG
    {
        name = Castor-1
        maxThrust = 286
        heatProduction = 100
        curveResource = PSPC
        chamberNominalTemp = 1500
        maxEngineTemp = 1745
        powerEffectName = Solid-Lower
        PROPELLANT
        {
            name = PSPC
            ratio = 1
            DrawGauge = True
        }
        atmosphereCurve
        {
            key = 0 247
            key = 1 232
        }
        thrustCurve
        {
            key = 0 0.2 0.6566457 0.6566457
            key = 0.08 0.7826234 11.38028 11.38028
            key = 0.12 1.091428 0 0
            key = 0.177 1.072238 -0.3012535 -0.3012535
            key = 0.832 0.9977316 0.2489114 0.2489114
            key = 0.91 1.052388 1.565168 1.565168
            key = 0.9730649 1.160696 0 0
            key = 1 0.93 -4.794868 -6.284284
        }
    }
}
```

Any reason why RF reports the curves as being null? Apart from that message the curves seem to work as intended.

[ModuleEnginesRF.cs check](https://github.com/NathanKell/ModularFuelSystem/blob/34c1bab161057730ef68dd5adae9d7375e756eee/Source/Engines/ModuleEnginesRF.cs#L175-L187)

[Castor 1 RO global engine config](https://github.com/KSP-RO/RealismOverhaul/blob/master/GameData/RealismOverhaul/Engine_Configs/Castor_1_Config.cfg)

**Comment by Starwaster** [comment]
Looks screwy. Lines 185-186 should be moved to before the null check takes place...

**Comment by PhineasFreak** [comment]
Hmm, that specific code did not change between releases. Still, this message never appears in previous RF versions.

Also, since the config already defines a thrust curve, why RF thinks that no curve is defined? That part of the code will not run if the thrust curve is not null.

**Comment by blowfishpro** [comment]
How does that code even get executed?  `thrustCurve` is initialized at the top of that method if it's null.  I feel like I'm missing something.

**Comment by Starwaster** [comment]
Regardless of whether the RF code has changed, don't forget that we've had substantial KSP changes this last update, and prior to that, Unity changes.

A newly constructed empty AnimationCurve (which is what a FloatCurve is a wrapper for) might not actually be considered an empty curve anymore. Maybe it's returning a null value on the curve since it was constructed with no keys or values...

That might not make a lot of sense but I'm having trouble seeing how else it could happen given that it's hitting a null check not once but TWICE. Only the second one actually tries to load the thrustCurve from the node. I'm tired though so I'll look at it again in the morning after coffee. Maybe I'm missing something too.

**Comment by jwvanderbeck** [comment]
It is entirely possible it simply failed to create the object the first time (maybe even the second time).  new blah doesn't GUARANTEE you get blah.  It can fail to allocate.

**Comment by blowfishpro** [comment]
@jwvanderbeck Is there any way that could happen and not throw an exception though?  There's nothing special about the `FloatCurve` class...

**Comment by blowfishpro** [comment]
Did a bit of digging here, apparently it's becoming null after `base.OnLoad()` https://github.com/NathanKell/ModularFuelSystem/blob/34c1bab161057730ef68dd5adae9d7375e756eee/Source/Engines/ModuleEnginesRF.cs#L168

Which is very strange because nothing in any of the superclasses seems to be modifying the value of `thrustCurve` and also because `OnLoad` doesn't actually try to load any of the fields (it's used for custom loading stuff, `Load` loads the field and calls `OnLoad`)

I'm at a bit of a loss here, any insight would be appreciated!

**Comment by blowfishpro** [comment]
Never mind, figured it out.  Here:

https://github.com/NathanKell/ModularFuelSystem/blob/34c1bab161057730ef68dd5adae9d7375e756eee/Source/Engines/ModuleEnginesRF.cs#L109

The thrust curve is nulled out if `useThrustCurve` is false.  It looks like a lot of engines define a `thrustCurve` but don't set `useThrustCurve = true`.  Few options here:

1. Require engines to define `useThrustCurve = true` if they're using a thrust curve, as in stock
2. Remove that code in `CreateEngine` that nulls it out.
3. Just get rid of the null warning

I'm inclined to go with 2, particularly since it doesn't require a bunch of configs to be changed.

**Comment by PhineasFreak** [comment]
@blowfishpro will anything break if the null check is not there? If the answer is yes then it would be better to just bite the bullet and add the required config option to the engine configs. Otherwise, removing the offending code will be easier than finding the configs that are missing the thrust curve flag.

**Comment by blowfishpro** [comment]
It's only becoming null because it's being set to null in `CreateEngine` (tested this, the error never shows up if this is removed).

Is there any benefit to requiring `useThrustCurve`?  The correct value for that is already determined here based on the presence of the curve node, so its original purpose in the stock module is no longer relevant.

**Comment by PhineasFreak** [comment]
Then, removing the check would be the better option.

**Comment by Starwaster** [comment]
Because the curve was initialized new and probably literally has a null value

Only after the null check are those configs loaded

**Comment by blowfishpro** [comment]
@Starwaster I figured out the problem and that's not it.  In C#, a reference will not be null after initializing it to a new object.  It is becoming null because code that sets it to null is being run.  To be sure of this, I inserted another null check after it is initialized but before `OnLoad` is called, and it passed every time.
